pkg/deploy/deployer.go

```diff
@@ -146,3 +151,63 @@ func (d *deployer) getParameters(ps map[string]interface{}) *arm.Parameters {
 
 	return parameters
 }
+
+func (d *deployer) deploy(ctx context.Context, t map[string]interface{}, p *arm.Parameters, rgName, deploymentName, vmssName string) (err error) {
+	for i := 0; i < 3; i++ {
+		d.log.Printf("deploying %s", deploymentName)
+		err = d.deployments.CreateOrUpdateAndWait(ctx, rgName, deploymentName, mgmtfeatures.Deployment{
+			Properties: &mgmtfeatures.DeploymentProperties{
+				Template:   t,
+				Mode:       mgmtfeatures.Incremental,
+				Parameters: p.Parameters,
+			},
+		})
+		if serviceErr, ok := err.(*azure.ServiceError); ok &&
+			serviceErr.Code == "DeploymentFailed" &&
+			i < 1 {
+			// on new RP deployments, we get a spurious DeploymentFailed error
+			// from the Microsoft.Insights/metricAlerts resources indicating
+			// that rp-lb can't be found, even though it exists and the
+			// resources correctly have a dependsOn stanza referring to it.
+			// Retry once.
+			d.log.Print(err)
+			continue
+		}
+		if err != nil {
+			if deleted := d.removeFailedScaleset(ctx, rgName, vmssName); deleted && i < 2 {
+				continue // Retry deployment after deleting failed VMSS.
+			}
+			return // when i == 2, on final attempt
+		}
+		break
+	}
+	return
+}
+
+func (d *deployer) removeFailedScaleset(ctx context.Context, rgName, vmssName string) (deleted bool) {
+	// Check if scaleset exists. If not, no need to delete it.
+	vmss, err := d.vmss.Get(ctx, rgName, vmssName)
+	if isNotFound(err) {
+		return true
+	}
+
+	// If it exists and is in failed state, try deleting it so naming conflict doesn't occur during retry.
+	if *vmss.ProvisioningState == string(mgmtcompute.ProvisioningStateFailed) {
+		d.log.Printf("deleting failed scaleset %s", *vmss.Name)
+		err = d.vmss.DeleteAndWait(ctx, rgName, *vmss.Name)
+		if err != nil {
+			d.log.Print(err)
+			return false
+		}
+		return true
+	}
+
+	return false
```

Comment:
We need true here, I think?

Reply:
If the VMSS exists and was not in Failed provisioning state, we return `false` here, since false means "not deleted". At this point in the code, we would not have deleted it.